app/controllers/published_pages_controller.rb

```diff
@@ -5,14 +5,30 @@ class PublishedPagesController < ApplicationController
   skip_before_action :preload_json
   skip_before_action :check_xhr, :verify_authenticity_token, only: [:show]
   before_action :ensure_publish_enabled
+  before_action :redirect_to_login_if_required, except: [:show]
 
   def show
     params.require(:slug)
 
     pp = PublishedPage.find_by(slug: params[:slug])
     raise Discourse::NotFound unless pp
 
-    guardian.ensure_can_see!(pp.topic)
+    return if enforce_login_required!
+
+    if !pp.public
```

Comment:
Indeed will look at implementing this

Reply:
After trying to implement this, I consider it kinda complicated given we have to handle the /login logic and the custom exception to avoid including ember. I will postpone this for now, and think of a better solution.